torch/csrc/jit/backends/backend_debug_handler.h

```diff
@@ -117,16 +119,72 @@ namespace jit {
  *  [M.forward, source range] -> [aten::mul's source range] We need to track
  *  mul's source range and inlined CS both.
  */
+using DebugHandleType = int64_t;
 class TORCH_API BackendDebugHandleManager {
  public:
   BackendDebugHandleManager() = default;
   int64_t getNextDebugHandleForInlinedCallStackPtr(const Node* node);
-  std::unordered_map<int64_t, DebugInfoPair> getCallStackPtrMap();
+  std::unordered_map<DebugHandleType, DebugInfoPair> getCallStackPtrMap();
+
+ private:
+  static std::atomic<DebugHandleType> unique_debug_handle_;
+  std::unordered_map<DebugHandleType, DebugInfoPair>
+      handles_to_inlined_callstack_ptrs_;
+};
+
+// Copied from torch/csrc/jit/api/object.h
+// to avoid including that file.
+using ObjectPtr = c10::intrusive_ptr<c10::ivalue::Object>;
+using DelegateDebugInfoMapType =
+    std::unordered_map<DebugHandleType, DebugInfoPair>;
+
+/*
+ * This class is used to generate debug info map (module's inline callstack ptr
+ * map). It is initialized with module's object ptr and during ctor it will
+ * instantiate debug_handle_manager and initialize thread local pointer to it.
+ * backend's preprocess will call generate_debug_handles, which uses
+ * debug_handle_manager to generate debug handles. When lowering process
+ * finishes, calling stopRecording will extract debug info map from
+ * debug_handle_manager and save it in an static global map to be
+ * used/serialized later.
+ */
+class BackendModuleDebugInfoRecorder {
+ public:
+  BackendModuleDebugInfoRecorder(ObjectPtr module_ptr);
+  ~BackendModuleDebugInfoRecorder();
+  // Reason this is not done as RAII is that work done in stopRecording
+  // can throw, and throwing with dtor will call terminate and thus voids any
+  // exception catching at a higher level.
+  void stopRecording();
+
+ private:
+  BackendDebugHandleManager debug_handle_manager;
+  ObjectPtr module_ptr_;
+};
+
+BackendDebugHandleManager* getBackendDebugHandleManager();
```

Comment:
That sounds good. I agree that 1) the disjoint nature of this is not exactly nice and 2) more importantly passing instance of the debug info recorder via preprocess forces you to consider the debug info when a new backend introduces its own preprocess.

Reply:
BTW in the meantime I will simplify BackendModuleDebugInfoRecorder since it is just a wrapper around BackendDebugHandleManager. I will merge BackendDebugHandleManager into the former. We still will have to write RAII that initializes thread local pointer of type BackendModuleDebugInfoRecorder, as opposed to of type BackendDebugHandleManager before. Then when we make the changes you suggested, we can remove RAII class and thread local stuff.

Reply:
Sounds good to me too, to put tightly couple components together. 

Currently, BackendPreprocessFunction has the signature of (module, method_compile_spec).
When passing an instance of BackendModuleDebugInfoRecorder to BackendPreprocessFunction, could we piggyback it to method_compile_spec? Something like 
`
["Debug_recorder": IValue<custom_class>(BackendModuleDebugInfoRecorder)].
`
By doing this,
1. We don't change the schema of BackendPreprocessFunction
2. BackendModuleDebugInfoRecorder could be optional, which is the intention?